pkg/graveler/graveler.go

```diff
@@ -1903,16 +1905,135 @@ func (g *KVGraveler) Reset(ctx context.Context, repositoryID RepositoryID, branc
 		return err
 	}
 
-	g.dropTokens(ctx, tokensToDrop)
+	g.dropTokens(ctx, tokensToDrop...)
+	return nil
+}
+
+// resetKey resets given key on branch
+// Since we cannot (will not) modify sealed tokens data, we overwrite changes done on entry on a new staging token, effectively reverting it
+// to the current state in the branch committed data. If entry is not committed return an error
+func (g *KVGraveler) resetKey(ctx context.Context, repositoryID RepositoryID, branch *Branch, key Key, stagedValue *Value, st StagingToken) error {
+	isCommitted := true
+	committed, err := g.Get(ctx, repositoryID, branch.CommitID.Ref(), key)
+	if err != nil {
+		if !errors.Is(err, ErrNotFound) {
+			return err
+		}
+		isCommitted = false
+	}
+
+	if isCommitted { // entry committed and changed in staging area => override with entry from commit
+		if reflect.DeepEqual(committed, stagedValue) {
+			return nil // No change
+		}
+		if err = g.StagingManager.Set(ctx, st, key, committed, true); err != nil {
+			return err
+		}
+	} else if !isCommitted && stagedValue != nil { // entry not committed and changed in staging area => override with tombstone
+		if err = g.StagingManager.Set(ctx, st, key, nil, true); err != nil {
+			return err
+		} // If not committed and staging == tombstone => ignore
+	}
 	return nil
 }
 
 func (g *KVGraveler) ResetKey(ctx context.Context, repositoryID RepositoryID, branchID BranchID, key Key) error {
-	return g.db.ResetKey(ctx, repositoryID, branchID, key)
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_STAGING_WRITE)
+	if err != nil {
+		return err
+	}
+	if isProtected {
+		return ErrWriteToProtectedBranch
+	}
+	// New sealed tokens list after change includes current staging token
+	newSealedTokens := make([]StagingToken, 0)
+	newStagingToken := generateStagingToken(repositoryID, branchID)
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		newSealedTokens = []StagingToken{branch.StagingToken}
```

Comment:
Let's add a compaction task, this implementation makes it necessary.

Reply:
https://github.com/treeverse/lakeFS/issues/3780